### ISSUE TITLE
Add bucket sampler

### ISSUE THREAD

**Issue by erip** [pytorch/text#1474]
## 🚀 Feature

**Motivation**

The legacy `BucketIterator` was convenient because it could batch samples by length to minimize padding. It had many disadvantages because of its API and non-comformance with other parts of the pytorch data{sets,loader} ecosystem. It would be nice if torchtext supported the spirit of the `BucketIterator` by way of a `Sampler`.

**Pitch**

A sampler with the ability to specify maximum bucket size should be added similar to those in torchnlp and allennlp. This can be used with existing datasets in torchtext but as a kwarg to the pytorch `DataLoader` so sampling minimizes padding.

**Alternatives**

Users who want this functionality need to implement their own samplers.

**Additional context**

N/A

**Comment by parmeet** [comment]
Thanks @erip for raising this issue. This indeed is one of the important issues to be resolved and we understand that many users rely on this feature. We are currently working on updating torchtext datasets to build on [torchdata DataPipes](https://github.com/pytorch/data). Fortunately, the equivalent functionality is already implemented [here](https://github.com/pytorch/data/blob/main/torchdata/datapipes/iter/transform/bucketbatcher.py), hence users should be able to sample batches with similar lengths out-of-the-box. 

We will add a tutorial to demonstrate this functionality. Let me know if you would be interested in making this contribution and I can help with the code pointers :)

**Comment by erip** [comment]
I'd definitely be interested in contributing!

**Comment by erip** [comment]
As an aside, I've also got a [max token sampler](https://gist.github.com/erip/81d2816f71ba2e95668095e5a1e1040e) which is a bit different, but may be of interest. Not sure if it makes sense to include it in torchtext, but if you think it's worth a discussion I'd be happy to contribute it.

**Comment by erip** [comment]
@nateanl thanks for the pointer! I think it looks pretty good, but a couple of questions:

- it looks like shuffling happens unconditionally. Does it make sense to add a shuffle kwarg to the sampler?
- I didn't realize Datasets had a len_list attr and I can't find that in the pytorch source. Is that populated somewhere in torchaudio?
- one neat trick to promote diversity in batches is to add some small random noise to the lengths of samples so you don't see the same examples each epoch. My gist has an impl of that -- WDYT about adding an optional feature for that in this notional upstream sampler?

**Comment by nateanl** [comment]
The `len_list` is in the custom HuBERT dataset, I think your impl is better, we should assume the dataset is unsorted and sort the indices by passing the unsorted lengths list.

I'm not sure shuffle the samples in the batch will improve the training, will the gradient differ by shuffling? What I did is shuffle the samples in the same bucket so that you can get different samples every epoch. I don't shuffle the order of buckets because that may introduce a large length gap in the same batch, when transferring to a new bucket.

In terms of creating the buckets, I use `num_buckets ` as the argument and set the boundaries by ``interval = (max_len - min_len) // num_buckets``. There are other ways to do it, for example, set `boundaries` as the argument. Which one do you think is better for use?

**Comment by nateanl** [comment]
```python
random.shuffle(buckets[k])
```
Yeah this shuffling will generate different batches, assume the bucket size is greater than `batch_size`, but shuffle the samples in the same batch may not change the gradient, what do you think ;)

`max_len` and `min_len` sounds good to me. Does it mean the samples whose length is `< min_len` or `> max_len` will be filtered out? If so, we can add such filtering to the new DataPipes code.

**Comment by erip** [comment]
> Yeah this shuffling will generate different batches

Ah, I misread the code -- this makes total sense. I agree that this will have more impact than shuffling within a batch. I still think it is good to give an option to the user about whether there should be shuffling at all. maybe just `shuffle: bool = True` by default?

> Does it mean the samples whose length 

Yes, exactly. See [here](https://gist.github.com/erip/81d2816f71ba2e95668095e5a1e1040e#file-max_token_per_batch_sampler-py-L25-L30). There's probably some error checking to add here (what if `lengths` is the empty list after filtering? 🙀 ), but otherwise seems OK.
